service/src/test/java/com/commercetools/util/ClientConfigurationUtilTest.java

```diff
@@ -0,0 +1,36 @@
+package com.commercetools.util;
+
+import com.commercetools.pspadapter.tenant.TenantConfig;
+
+import io.sphere.sdk.client.BlockingSphereClient;
+import io.sphere.sdk.client.SphereClient;
+import io.sphere.sdk.client.SphereClientConfig;
+import org.junit.Before;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.mockito.Mock;
+import org.mockito.runners.MockitoJUnitRunner;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.mockito.Mockito.when;
+
+@RunWith(MockitoJUnitRunner.class)
+public class ClientConfigurationUtilTest {
+
+    @Mock
+    private TenantConfig tenantConfig;
+
+    @Before
+    public void setUp() throws Exception {
+      when(tenantConfig.getSphereClientConfig())
+        .thenReturn(SphereClientConfig.of("test-key", "test-client-id", "test-client-secret"));
+    }
+
+    @Test
+    public void executeCreateClient_isBlocking() {
+        SphereClient sphereClient = ClientConfigurationUtil.createClient(tenantConfig.getSphereClientConfig());
+        assertThat(sphereClient instanceof BlockingSphereClient).isTrue();
```

Comment:
This does not really ensure that retry decorator is also included in the chain?

Reply:
I tried. However, after the chain, retry decorator is wrapped inside blocking decorator in JVM SDK. There is no getter method to obtain the previous one from decorator returned.

Reply:
The lowest underlying HTTP client is instantiated like this
`return SphereClientFactory.of().createClient(clientConfig)`
https://github.com/commercetools/commercetools-payone-integration/blob/101635cd40611409e9cfcfb0e8172ca0028041fc/service/src/main/java/com/commercetools/util/GenericClientConfigurationUtil.java#L12

After discussed with Jude, to mock this static method, our approach is to adapt Powermock plugins on top of Mockito. The unit test case implemented as below
https://github.com/commercetools/commercetools-payone-integration/blob/101635cd40611409e9cfcfb0e8172ca0028041fc/service/src/test/java/com/commercetools/util/ClientConfigurationUtilTest.java#L52

However, even adapting Powermock, final class is not allowed to be mocked.
https://travis-ci.org/github/commercetools/commercetools-payone-integration/builds/720694564

Reply:
Discussed with jdk-svm team. They suggested to change the implementation in this way, so that `decorateClient` can be tested independently in unit test, while 5xx http response can be directly assign to underlying `SphereClient` instead of mocking it.

```
public static BlockingSphereClient createClient(@Nonnull final SphereClientConfig clientConfig) {
        final SphereClient underlyingClient = SphereClientFactory.of().createClient(clientConfig);
        return decorateClient(underlyingClient)
    }
public static BlockingSphereClient decorateClient(final SphereClient client) {
        final SphereClient retryClient = withRetry(client);
        final SphereClient limitedClient = withLimitedParallelRequests(retryClient);
        return withBlocking(limitedClient);
    }
}
```